src/A2A/Models/A2AResponse.cs

```diff
@@ -1,21 +1,131 @@
-[JsonDerivedType(typeof(TaskArtifactUpdateEvent), "artifact-update")]
-[JsonDerivedType(typeof(AgentMessage), "message")]
-[JsonDerivedType(typeof(AgentTask), "task")]
-public abstract class A2AEvent;
+/// <param name="kind">The <c>kind</c> discriminator value</param>
+[JsonConverter(typeof(A2AEventConverterViaKindDiscriminator<A2AEvent>))]
+[JsonDerivedType(typeof(TaskStatusUpdateEvent))]
+[JsonDerivedType(typeof(TaskArtifactUpdateEvent))]
+[JsonDerivedType(typeof(AgentMessage))]
+[JsonDerivedType(typeof(AgentTask))]
+// You might be wondering why we don't use JsonPolymorphic here. The reason is that it automatically throws a NotSupportedException if the 
+// discriminator isn't present or accounted for. In the case of A2A, we want to throw a more specific A2AException with an error code, so
+// we implement our own converter to handle that, with the discriminator logic implemented by-hand.
+public abstract class A2AEvent(A2AEventKind kind)
+{
+    /// <summary>
+    /// The 'kind' discriminator value
+    /// </summary>
+    [JsonRequired, JsonPropertyName(A2AEventConverterViaKindDiscriminator<A2AEvent>.DiscriminatorPropertyName), JsonInclude, JsonPropertyOrder(int.MinValue)]
+    public A2AEventKind Kind { get; internal set; } = kind;
+}
 
 /// <summary>
 /// A2A response objects.
 /// </summary>
-[JsonPolymorphic(TypeDiscriminatorPropertyName = "kind")]
-[JsonDerivedType(typeof(AgentMessage), "message")]
-[JsonDerivedType(typeof(AgentTask), "task")]
-public abstract class A2AResponse : A2AEvent;
\ No newline at end of file
+/// <param name="kind">The <c>kind</c> discriminator value</param>
+[JsonConverter(typeof(A2AEventConverterViaKindDiscriminator<A2AResponse>))]
+[JsonDerivedType(typeof(AgentMessage))]
+[JsonDerivedType(typeof(AgentTask))]
+// You might be wondering why we don't use JsonPolymorphic here. The reason is that it automatically throws a NotSupportedException if the 
+// discriminator isn't present or accounted for. In the case of A2A, we want to throw a more specific A2AException with an error code, so
+// we implement our own converter to handle that, with the discriminator logic implemented by-hand.
+public abstract class A2AResponse(A2AEventKind kind) : A2AEvent(kind);
+
+internal class A2AEventConverterViaKindDiscriminator<T> : JsonConverter<T> where T : A2AEvent
+{
+    internal const string DiscriminatorPropertyName = "kind";
+
+    public override T Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
+    {
+        using var document = JsonDocument.ParseValue(ref reader);
+        var root = document.RootElement;
+
+        if (!root.TryGetProperty(DiscriminatorPropertyName, out var kindProp) || kindProp.ValueKind is not JsonValueKind.String)
+        {
+            throw new A2AException($"Missing required '{DiscriminatorPropertyName}' discriminator for {typeof(T).Name}.", A2AErrorCode.InvalidRequest);
+        }
+
+        T? evt = null;
```

Comment:
Not sure what is the project's convention, but I personally would avoid using abbreviations for better code readability.